forge-game/src/main/java/forge/game/trigger/Trigger.java

```diff
@@ -623,4 +625,34 @@ public void setOverridingAbility(SpellAbility overridingAbility0) {
         super.setOverridingAbility(overridingAbility0);
         overridingAbility0.setTrigger(this);
     }
+
+    boolean whileKeywordCheck(final String keyword, final Map<AbilityKey, Object> runParams) {
+        SpellAbility sa;
+
+        IndividualCostPaymentInstance currentPayment =
+                (IndividualCostPaymentInstance) runParams.get(AbilityKey.IndividualCostPaymentInstance);
+        if (currentPayment != null) {
+            sa = currentPayment.getPayment().getAbility();
+            if (sa != null) {
+                if (whileKeywordSACheck(keyword, sa)) return true;
+            }
+        }
+
+        CostPaymentStack stack = (CostPaymentStack) runParams.get(AbilityKey.CostStack);
+        for (IndividualCostPaymentInstance individual : stack) {
+                sa = individual.getPayment().getAbility();
+                if (whileKeywordSACheck(keyword, sa)) return true;
+        }
+
+        return false;
+    }
+
+    private boolean whileKeywordSACheck(final String keyword, final SpellAbility sa) {
+        if (sa == null || sa.getHostCard() == null) return false;
+        if (sa.isAbility()) {
+            if (keyword.equals("Craft") && sa.isCraft()) return true;
+        }
+        return sa.isSpell() && sa.getHostCard().hasStartOfUnHiddenKeyword(keyword);
```

Comment:
it still would trigger if cast a spell with Craft and exiling it for Mana

Reply:
OK, I think I have this wired up now but Foul Emissary is broken – help!? I need to sleep 🛏️ 

On the Emissary, I tried a few different values for `WithKeyword$` but the TriggerSacrificed.performTest() is not even being run. 😩 I even tried removing WithKeyword param altogether for a test and sacrificed Foul Emissary to Wretched Gryff... no breakpoint in performTest() and no Trigger

Reply:
what exactly did you do ? it works for me, or have you fixed this already?

Reply:
tested with JDK 21

it failed the moment i tried to cheat the Foul Emissary into play instead of casting it, because LastStateBattlefield wasn't set